c++/src/kj/compat/http.c++

```diff
@@ -4093,6 +4094,85 @@ private:
   kj::ArrayPtr<byte> leftover;
 };
 
+class AsyncIoStreamWithDeferredInput final: public kj::AsyncIoStream {
+  // This AsyncIoStream immediately allows write to the given wrapped AsyncIoStream
+  // but will delay reads until after a guard promise is resolved. When the guard
+  // promise resolves, it may provide a released buffer that will be read out first.
+  // The primary use case for this impl is to support pipelined CONNECT calls which
+  // optimistically allow outbound writes to happen while establishing the CONNECT
+  // tunnel has not yet been completed. If the guard promise rejects, the stream
+  // is permanently errored and existing pending calls (reads and writes) are canceled.
+public:
+  AsyncIoStreamWithDeferredInput(
+      kj::Own<kj::AsyncIoStream> inner,
+      kj::Promise<kj::Maybe<HttpInputStreamImpl::ReleasedBuffer>> guard)
+      : inner(kj::mv(inner)),
+        guard(handleGuard(kj::mv(guard))) {}
+
+  // AsyncInputStream
+  Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+    if (guardReleased) {
+      return inner->tryRead(buffer, minBytes, maxBytes);
+    }
+    return guard.addBranch().then([this, buffer, minBytes, maxBytes] {
+      return inner->tryRead(buffer, minBytes, maxBytes);
+    });
+  }
+
+  Maybe<uint64_t> tryGetLength() override {
+    return nullptr;
+  }
+
+  kj::Promise<uint64_t> pumpTo(AsyncOutputStream& output, uint64_t amount) override {
+    if (guardReleased) {
+      return inner->pumpTo(output, amount);
+    }
+    return guard.addBranch().then([this, &output, amount] {
+      return inner->pumpTo(output, amount);
+    });
+  }
+
+  // AsyncOutputStream
+
+  void shutdownWrite() override {
+    inner->shutdownWrite();
+  }
+
+  kj::Maybe<kj::Promise<uint64_t>> tryPumpFrom(AsyncInputStream& input, uint64_t amount) override {
+    return input.pumpTo(*inner, amount);
+  }
+
+  Promise<void> write(const void* buffer, size_t size) override {
+    return inner->write(buffer, size);
+  }
+
+  Promise<void> write(ArrayPtr<const ArrayPtr<const byte>> pieces) override {
+    return inner->write(pieces);
+  }
+
+  Promise<void> whenWriteDisconnected() override {
+    return inner->whenWriteDisconnected();
+  }
+
+private:
+  kj::Own<kj::AsyncIoStream> inner;
+  kj::ForkedPromise<void> guard;
+  bool guardReleased = false;
+
+  kj::ForkedPromise<void> handleGuard(
+      kj::Promise<kj::Maybe<HttpInputStreamImpl::ReleasedBuffer>> guard) {
+    return guard.then([this](kj::Maybe<HttpInputStreamImpl::ReleasedBuffer> buffer) mutable {
+      guardReleased = true;
+      KJ_IF_MAYBE(b, buffer) {
+        // We only need to replace the inner stream if a buffer is provided.
```

Comment:
I suppose you could also check if b->leftover is empty, and if so, don't bother adding the wrapper. This is probably a common case with CONNECT: the headers arrive before the initial data, so there's no buffer.

Reply:
Ah right, yeah good idea